awx/main/models/credential/__init__.py

```diff
@@ -701,527 +720,525 @@ def build_extra_vars_file(vars, private_dir):
-                'type': 'string',
-                'secret': True,
-                'help_text': ugettext_noop('Security Token Service (STS) is a web service '
-                                           'that enables you to request temporary, '
-                                           'limited-privilege credentials for AWS Identity '
-                                           'and Access Management (IAM) users.'),
-            }],
-            'required': ['username', 'password']
-        }
-    )
+class NativeCredentialType(SimpleNamespace):
+
+    registry = {}
+
+    def __init__(self, namespace, **kwargs):
+        for k in ('inputs', 'injectors'):
+            if k not in kwargs:
+                kwargs[k] = {}
+        super(NativeCredentialType, self).__init__(namespace=namespace, **kwargs)
+        if namespace in NativeCredentialType.registry:
+            raise ValueError(
+                'a NativeCredentialType with namespace={} is already defined'.format(namespace)
+            )
+        NativeCredentialType.registry[namespace] = self
+
+    def create(self):
+        return CredentialType(
+            native_source=self.namespace,
+            kind=self.kind,
+            name=self.name,
+            managed_by_tower=True,
+            inputs=self.inputs,
+            injectors=self.injectors,
+        )
 
 
-@CredentialType.default
-def openstack(cls):
-    return cls(
-        kind='cloud',
-        name=ugettext_noop('OpenStack'),
-        managed_by_tower=True,
-        inputs={
-            'fields': [{
-                'id': 'username',
-                'label': ugettext_noop('Username'),
-                'type': 'string'
-            }, {
-                'id': 'password',
-                'label': ugettext_noop('Password (API Key)'),
-                'type': 'string',
-                'secret': True,
-            }, {
-                'id': 'host',
-                'label': ugettext_noop('Host (Authentication URL)'),
-                'type': 'string',
-                'help_text': ugettext_noop('The host to authenticate with.  For example, '
-                                           'https://openstack.business.com/v2.0/')
-            }, {
-                'id': 'project',
-                'label': ugettext_noop('Project (Tenant Name)'),
-                'type': 'string',
-            }, {
-                'id': 'domain',
-                'label': ugettext_noop('Domain Name'),
-                'type': 'string',
-                'help_text': ugettext_noop('OpenStack domains define administrative boundaries. '
-                                           'It is only needed for Keystone v3 authentication '
-                                           'URLs. Refer to Ansible Tower documentation for '
-                                           'common scenarios.')
-            }],
-            'required': ['username', 'password', 'host', 'project']
+NativeCredentialType(
```

Comment:
@jakemcdermott your PR that does this: https://github.com/ansible/awx/pull/3098/files#diff-8923e305beef7163bf0fcb6d87998376R583

...would be updated to invoke _this_ instead; each plugin would need to define a (unique) `namespace` for registration, and in this way, we'd have a per-plugin `main_credentialtype` row w/ a `native_source` value pointed at the namespace for the plugin (so we knew which code to load to determine the inputs and injectors).

Reply:
e.g.,

```python
@classmethod
def load_plugin(cls, plugin):
    NativeCredentialType(
        namespace=plugin.namespace,
        name=plugin.name,
        kind='external',
        inputs=plugin.inputs
    )
```

Reply:
This makes sense - this will store a reference (namespace) in the db to the latest set of inputs. 👍